### DISCUSSION TITLE
How to create tmp output dynamically?

### DISCUSSION THREAD

**Discussion by rkight66** [nextgenhealthcare/connect#5528] [Q&A] [answered]
How would you create the tmp output dynamically?
Can I build a tmp output in a channel without having an actual template in the output template window?
I'm on version 3.8.1 and building C-CDA/CCDs and with all the different ways the message can be created, I'm wondering if the tmp can be built on the fly instead of using a standard template that won't fit every situation.

Thanks in advance for any and all help.

**Comment by pacmano1** [comment, marked as answer]
You don't need anything in the outbound message template (e.g. if the inbound message is XML, since you can just edit msg).

If you are transforming from another data type, just be a bare minimum stub in there e.g. `<xml></xml>` and blow it away as needed.

**Reply by rkight66** [reply]
So, would you create all the CCD sections and add based on the XPath?
Example:
My minimum skeleton:
<?xml-stylesheet type='text/xsl' href='CDA.xsl' ?>
<ClinicalDocument xmlns="urn:hl7-org:v3" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
</ClinicalDocument>


Say I wanted to drop in the first block of header data like this:
<realmCode code="US"></realmCode>
<typeId extension="TSTDOC_HD00050" root="2.16.840.1.113883.1.3"></typeId>
<templateId root="2.16.840.1.113883.10.20.22.1.2" extension="2015-08-01" />
<code code="31433-9" codeSystem="2.16.840.1.113883.6.1" codeSystemName="LOINC" displayName="Summarization of Episode Note" />
<title>Clinical Summary</title>
<effectiveTime value="20221114080000-0500" />
<confidentialityCode code="N" codeSystem="2.16.840.1.113883.5.25" displayName="Normal" />
<languageCode code="US-en" />


The <recordTarget> would be next.

How would I reference the new data and add it to the skeleton?  Just add to the XPath?
I'm not sure how this would go. I'm used to dragging a path from the output template.

Thanks in advance.

**Reply by siddharthsharma9258** [reply]
Something on the lines of `tmp['title']['abbreviation'] = "Mr."` will append an abbreviation XML node under title node.

**Reply by rkight66** [reply]
For full disclosure, this is my first time working on CCDs. I have been attempting this effort with a full sample CCD as my outbound template. I'm using JSON payloads to pull the data to map into the CCD data elements, like the allergies, medications, etc. The trouble I'm having is every time I attempt to access items within the component|structureBody|component areas called the section, I get an error stating "Cannot read property SECTION from undefined".

I have done some testing to put values of the levels above the SECTION into channel variables and they are all coming back as undefined. 
It seems that the areas of the CCD that are seen as arrays are inaccessible.   Like these:   tmp['component']['structureBody']['component'][3]['section']

Does anyone have any ideas on how to handle accessing these data elements? 

Thanks in advance.

**Reply by pacmano1** [reply]
Generating CDAs is very complex.  You will need to create any empty xml node if it does not exist.  You can't just do an assignment to a non existing node or array element.

**Reply by rkight66** [reply]
I tried to create the nodes in the Array as tmp['component']['structureBody']['component'] = [], but the nodes in the array are still undefined. Do I need to create each subsequent node?  Am I creating the nodes correctly? Or is there a routine to use?

**Reply by pacmano1** [reply]
Wander over to the mirth forums, there are lots of posts on working with XML.   Also read the e4x spec.

**Comment by tom08zehn** [comment]
As far as I remember, the `tmp`variable doesn't exist in a transformer script if you don't enter something in the output message template... of course you could define it (`var tmp = '';`) but then it still doesn't represent the transformed message.

Even for output data type RAW you need to enter at least one character (can be space), for output data type JSON you need to enter `{}` etc. If output data type is JSON and you enter `foo` then it cannot be parsed as JSON unless you overwrite it in your code. So it doesn't need to match the (stringified) data type but still, you need to enter at least something.
